common/middleware.go

```diff
@@ -133,6 +140,63 @@ func write400(writer http.ResponseWriter, m Marshaller, request string, err erro
 	writer.Write(outData)
 }
 
+func doHook(before bool, route Route, restContext RestContext, body string) (string, error) {
+	hook := route.Hook
+	if hook == nil {
+		log.Printf("doHook(): No hook for %s %s", route.Method, route.Pattern)
+		return "", nil
+	}
+	hookInfo := fmt.Sprintf("Hook for %s %s: %s", route.Method, route.Pattern, hook.Executable)
+	if before && strings.ToLower(hook.When) != "before" {
+		log.Printf("doHook(): We are in before and %s is for after", hookInfo)
+		return "", nil
+	}
+	if !before && strings.ToLower(hook.When) != "after" {
+		log.Printf("doHook(): We are in after and %s is for before", hookInfo)
+		return "", nil
+	}
+	pathVars := restContext.PathVariables
+	queryVars := restContext.QueryVariables
+	clArgs := make([]string, len(pathVars)+len(queryVars)+1)
+
+	clArgs[0] = fmt.Sprintf("%s=%s", HookExecutableBodyArgument, body)
+	i := 1
+	for k, v := range pathVars {
+		clArgs[i] = fmt.Sprintf("%s=%s", k, v)
+		i++
+	}
+
+	for k, v := range queryVars {
+		clArgs[i] = fmt.Sprintf("%s=%s", k, v)
+		i++
+	}
+
+	var writer io.Writer
+	var err error
+	writer = nil
+	if hook.Output != "" {
+		log.Printf("doHook(): Writing output of %s to %s", hookInfo, hook.Output)
+		writer, err = os.Create(hook.Output)
+		if err != nil {
+			return "", err
+		}
+	} else {
+		log.Printf("doHook(): No output specified for %s", hookInfo)
+	}
+	log.Printf("doHook(): Running hook %s with %s", hook.Executable, clArgs)
+	cmd := exec.Command(hook.Executable, clArgs...)
```

Comment:
You will notice that this is actually being tested - using bash scripts, not mocked here. So...